charmhelpers/contrib/openstack/utils.py

```diff
@@ -1303,6 +1305,55 @@ def is_unit_paused_set():
         return False
 
 
+def manage_payload_services(action, services=None, charm_func=None):
+    """Run an action against all services.
+
+    An optional charm_func() can be called that can either raise an
+    Exception or return non None, None to indicate that the function
+    didn't execute cleanly.
+
+    The signature for charm_func is:
+    charm_func() -> message: string
+
+    charm_func() is executed after any services are stopped, if supplied.
+
+    The services object can either be:
+      - None : no services were passed (an empty dict is returned)
+      - a list of strings
+      - A dictionary (optionally OrderedDict) {service_name: {'service': ..}}
+      - An array of [{'service': service_name, ...}, ...]
+
+    @param action: Action to run: pause, resume, start or stop
+    @param services: OPTIONAL see above
+    @param charm_func: function to run for custom charm pausing.
+    @returns Bool, [] Whether succesfull, Status messages
+    @raises Exception(message) on an error for action_fail().
+    """
+    actions = {
+        'pause': service_pause,
+        'resume': service_resume,
+        'start': service_start,
+        'stop': service_stop}
```

Comment:
For robustness, it's probably worth adding a check like below here, as otherwise the `rc = actions[action](service)` will throw a `KeyError` which is a bit more obtuse.

```python
    action = action.lower()
    if action not in actions.keys():
        raise RuntimeError("action: {} must be on of: {}"
                           .format(action, ', '.join(actions.keys())
```